File: GFG POTD/007-July-2024/009-three-sum-closest.cpp
// https://www.geeksforgeeks.org/problems/three-sum-closest/1




//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;
// } Driver Code Ends
// User function template for C++

// arr : given vector of elements
// target : given sum value
class Solution {
  public:
    int threeSumClosest(vector<int> arr, int target) {
        sort(arr.begin(), arr.end());
        
        int n          = arr.size();
        int closestSum = INT_MAX;
        int minDiff    = INT_MAX;
        
        for (int i = 0; i < n - 2; ++i) {
            int low = i + 1;
            int high = n - 1;
            
            while (low < high) {
                int currentSum = arr[i] + arr[low] + arr[high];
                int currentDiff = abs(currentSum - target);
                
                if (currentDiff < minDiff) {
                    minDiff = currentDiff;
                    closestSum = currentSum;
                } else if (currentDiff == minDiff) {
                    closestSum = max(closestSum, currentSum);
                }
                
                if (currentSum < target) {
                    ++low;
                } else {
                    --high;
                }
            }
        }
        
        return closestSum;
    }
};

//{ Driver Code Starts.
int main() {
    int t;
    cin >> t;
    while (t--) {
        cin.ignore();
        vector<int> arr;
        string input;
        getline(cin, input);
        stringstream ss(input);
        int number;
        while (ss >> number) {
            arr.push_back(number);
        }
        int key;
        cin >> key;
        // cin.ignore();
        // if (f)
        //     return 0;
        Solution ob;
        cout << ob.threeSumClosest(arr, key) << endl;
    }
    return 0;
}
// } Driver Code Ends

File: GFG POTD/007-July-2024/009-three-sum-closest.py
# https://www.geeksforgeeks.org/problems/three-sum-closest/1




#User function Template for python3
# arr    : list of integers denoting the elements of the array
# target : as specified in the problem statement
class Solution:
    # TC : O(N^2)
    # SC : O(1)
    def threeSumClosest(self, arr, target):
        arr.sort()
        n = len(arr)
        closest_sum = float('inf')
        min_diff = float('inf')
    
        for i in range(n - 2):
            low  = i + 1
            high = n - 1
    
            while low < high:
                current_sum = arr[i] + arr[low] + arr[high]
                current_diff = abs(current_sum - target)
    
                if current_diff < min_diff:
                    min_diff = current_diff
                    closest_sum = current_sum
                elif current_diff == min_diff:
                    closest_sum = max(closest_sum, current_sum)
    
                if current_sum < target:
                    low += 1
                else:
                    high -= 1
        return closest_sum
        
#{ 
 # Driver Code Starts
#Initial Template for Python 3
if __name__ == '__main__':
    t = int(input())

    for _ in range(t):
        A = list(map(int, input().strip().split()))
        k = int(input())
        ob = Solution()
        print(ob.threeSumClosest(A, k))

# } Driver Code Ends